client/reader_slice_of_slices_iterator.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"io"
+
+	"github.com/m3db/m3db/interfaces/m3db"
+	"github.com/m3db/m3db/network/server/tchannelthrift/thrift/gen-go/rpc"
+	xio "github.com/m3db/m3db/x/io"
+)
+
+type readerSliceOfSlicesIterator struct {
+	segments       []*rpc.Segments
+	readers        []io.Reader
+	readersIdx     int
+	segmentReaders []m3db.SegmentReader
+	idx            int
+	len            int
+	closed         bool
+	pool           readerSliceOfSlicesIteratorPool
+}
+
+func newReaderSliceOfSlicesIterator(
+	segments []*rpc.Segments,
+	pool readerSliceOfSlicesIteratorPool,
+) *readerSliceOfSlicesIterator {
+	it := &readerSliceOfSlicesIterator{}
+	it.Reset(segments)
+	return it
+}
+
+func (it *readerSliceOfSlicesIterator) Next() bool {
+	if !(it.idx+1 < it.len) {
+		return false
+	}
+	it.idx++
+	it.readers = it.readers[:0]
+	it.readersIdx = 0
+	return true
+}
+
+func (it *readerSliceOfSlicesIterator) Current() []io.Reader {
```

Comment:
That's quite possible, however without these semantics we would have to create an array of iterators on the heap each time and pool those.  This iterator is implemented as such to provide `[]io.Reader` access to individual entries of an an existing `[][]xio.SegmentReader` without any allocations.